path_rotate_root.go

```diff
@@ -0,0 +1,248 @@
+	return &framework.Path{
+		Pattern: "rotate-root",
+		Fields:  map[string]*framework.FieldSchema{
+			// None
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback:                    b.pathRotateRoot,
+				ForwardPerformanceSecondary: true,
+				ForwardPerformanceStandby:   true,
+			},
+		},
+
+		HelpSynopsis: "Attempt to rotate the root credentials used to communicate with Azure.",
+		HelpDescription: "This path will attempt to generate new root credentials for the user used to access and manipulate Azure.\n" +
+			"The new credentials will not be returned from this endpoint, nor the read config endpoint.",
+	}
+}
+
+func (b *azureSecretBackend) pathRotateRoot(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	expiration := time.Now().AddDate(0, 6, 0)
+	passCred, warnings, err := b.rotateRootCredentials(ctx, req.Storage, expiration)
+	if err != nil {
+		return nil, err
+	}
+
+	resultData := map[string]interface{}{
+		"secret_id": *passCred.KeyID,
+		"end_date":  *passCred.EndDate,
+	}
+	if passCred.DisplayName != nil && *passCred.DisplayName != "" {
+		resultData["display_name"] = *passCred.DisplayName
+	}
+
+	resp := &logical.Response{
+		Data:     resultData,
+		Warnings: warnings,
+	}
+
+	return resp, nil
+}
+
+func (b *azureSecretBackend) rotateRootCredentials(ctx context.Context, storage logical.Storage, expiration time.Time) (cred api.PasswordCredential, warnings []string, err error) {
+	cfg, err := b.getConfig(ctx, storage)
+	if err != nil {
+		return api.PasswordCredential{}, nil, err
+	}
+
+	client, err := b.getClient(ctx, storage)
+	if err != nil {
+		return api.PasswordCredential{}, nil, err
+	}
+
+	// We need to use List instead of Get here because we don't have the Object ID
+	// (which is different from the Application/Client ID)
+	apps, err := client.provider.ListApplications(ctx, fmt.Sprintf("appId eq '%s'", cfg.ClientID))
+	if err != nil {
+		return api.PasswordCredential{}, nil, err
+	}
+
+	if len(apps) == 0 {
+		return api.PasswordCredential{}, nil, fmt.Errorf("no application found")
+	}
+	if len(apps) > 1 {
+		return api.PasswordCredential{}, nil, fmt.Errorf("multiple applications found - double check your client_id")
+	}
+
+	app := apps[0]
+
+	// Get a list of the current passwords to delete if adding a new credential succeeds. This list will not
+	// include the ID of the new password created via AddApplicationPassword
+	credsToDelete := []string{}
+	for _, cred := range app.PasswordCredentials {
+		credsToDelete = append(credsToDelete, *cred.KeyID)
+	}
+
+	uniqueID, err := uuid.GenerateUUID()
+	if err != nil {
+		return api.PasswordCredential{}, nil, fmt.Errorf("failed to generate UUID: %w", err)
+	}
+
+	// This could have the same username customization logic put on it if we really wanted it here
+	passwordDisplayName := fmt.Sprintf("vault-plugin-secrets-azure-%s", uniqueID)
```

Comment:
Wondering if the prefix is a bit long and could simply just be `vault_%s`, similar to application passwords?